Source/WebCore/platform/graphics/avfoundation/AudioVideoRendererAVFObjC.mm

```diff
@@ -802,7 +804,65 @@ @interface AVSampleBufferDisplayLayer (Staging_100128644)
 
 bool AudioVideoRendererAVFObjC::shouldBePlaying() const
 {
-    return m_isPlaying && !seeking();
+    return m_isPlaying && !seeking() && allRenderersHaveAvailableSamples();
+}
+
+void AudioVideoRendererAVFObjC::setHasAvailableAudioSample(TrackIdentifier trackId, bool flag)
+{
+    if (flag && allRenderersHaveAvailableSamples())
+        return;
+
+    auto it = m_audioTracksMap.find(trackId);
+    if (it == m_audioTracksMap.end())
+        return;
+    auto& properties = it->value;
+    if (properties.hasAudibleSample == flag)
+        return;
+    ALWAYS_LOG(LOGIDENTIFIER, flag);
+    properties.hasAudibleSample = flag;
+
+    updateAllRenderersHaveAvailableSamples();
+}
+
+void AudioVideoRendererAVFObjC::updateAllRenderersHaveAvailableSamples()
+{
+    bool allRenderersHaveAvailableSamples = true;
+
+    do {
+        if (m_enabledVideoTrackId && !m_hasAvailableVideoFrame) {
+            allRenderersHaveAvailableSamples = false;
+            break;
+        }
+
+        for (auto& properties : m_audioTracksMap.values()) {
+            if (!properties.hasAudibleSample) {
+                allRenderersHaveAvailableSamples = false;
+                break;
+            }
+        }
+    } while (0);
```

Comment:
We tend to use lambdas, so that we can use `return true;` or a method.

Reply:
it's code copied from MediaPlayerPrivateMediaSourceAVFObjC